Add deploy with API details (#3779)
astronomer/docs: astro/best-practices/deploy-with-api.md

---
title: 'Deploy to Astro with the API'
sidebar_label: 'Deploy with API'
id: deploy-with-api
---

> While you can deploy code to Astro using the GitHub Integration, the Astro CLI, or by configuring a CI/CD pipeline, you might prefer to use the Astro API to automate deploying code. Using the Astro API can solve some of the deploy automation issues that can result because some CI/CD configurations are not compatible with the GitHub Integration and the Astro CLI requires you to install Docker.

Reviewer: ```suggestion
While you can deploy code using the Astro GitHub Integration, the Astro CLI, or by configuring a CI/CD pipeline, you might prefer to use the Astro API to automate deploying code. Using the Astro API is the most straightforward way to deploy code and requires very few additional dependencies, meaning it is compatible with most CI/CD configurations.
```

> While you can deploy code to Astro using the GitHub Integration, the Astro CLI, or by configuring a CI/CD pipeline, you might prefer to use the Astro API to automate deploying code. Using the Astro API can solve some of the deploy automation issues that can result because some CI/CD configurations are not compatible with the GitHub Integration and the Astro CLI requires you to install Docker.

Reviewer: We mention lack of Docker in the original edit as a benefit, but I see Docker commands all over the scripts and steps. Can we get clarification on what the actual benefit is here? I think customers are more interested in that you no longer have to install the Astro CLI itself, rather than Docker.

> Instead of using the other code deploy options, you can instead use the Astro API Deploy endpoints to perform `IMAGE_ONLY`, `DAG_ONLY`, or both `IMAGE_AND_DAG` Deploys.

Reviewer: ```suggestion
If the Astro API has access to your Astro project files, you can use the Astro API `deploy` endpoints to complete an image deploy or DAG deploy. You can then implement scripts to automate deploys as a workaround to not using the Astro CLI or Astro GitHub integration.
```
How does this code-level detail help the user understand the utility of the API? IMO we are surfacing this too soon - I tried to elaborate more on the benefit of the API and the general practice of deploying code using the API

> This guide contains an example of an entire deploy flow and the respective scripts that you can use to deploy code with the Astro API using bash scripts. It also includes an example of a script that can make deploys depending on the files changed, in case you want to use those steps and integrate into your own CI/CD pipelines.

Reviewer: ```suggestion
This best practice guide walks through the steps that are necessary to deploy code to Astro using the Astro API. It also provides example bash scripts that implement these steps for different types of deploy workflows in your own CI/CD pipelines.
```
Second sentence repetitive of first or am I missing a detail

## Feature overview

> This guide highlights the following Astro features to use for automating code deploys:

Reviewer: ```suggestion
This guide highlights the following Astro features:
```

- The Astro API [`Deploy` endpoint](https://www.astronomer.io/api/platform-api-reference/deploy/list-deploys) to create and manage code deploys to an Astro Deployment.

## Prerequisites

These use cases assume you have:

- Sufficient permissions to deploy code to your Astro Deployments

> - At least one Astro Deployment

Reviewer: Link to create a deployment docs?

> - The following values configured:

Reviewer: Why are we referring to these as the script environment variable values rather than the actual objects? E.g. `ORGANIZATION_ID` > Your Organization ID

    - `ORGANIZATION_ID`

>     - `DEPLOYMENT_ID`

Reviewer: We don't explain how to find these values in this doc, nor do we link to where to find that info. We should do at least one of those.

Reviewer: The configuration is in the script, so these don't have to be preconfigured.

    - `ASTRO_API_TOKEN` - See [Create an API token](https://www.astronomer.io/docs/astro/automation-authentication#step-1-create-an-api-token)

>     - `AIRFLOW_PROJECT_PATH` - The path where your Airflow project exists.

Reviewer: ```suggestion
    - `AIRFLOW_PROJECT_PATH` - The path where your Airflow project exists.
```
Rewrite to "Astro project" across the whole guide? And what if you're only deploying DAGS?

## With DAG-only deploy enabled

If you have DAG-only deploys enabled, you can create scripts for complete project deploys, DAG-only deploys, or image-only deploys.

### Project deploy

> 1. Create the `Deploy`

Reviewer: This step provides very little context for the actual action the user would be taking. Even if we don't provide a whole API example call, we should at least explain that this is the construction of an API call and also explain how to get those 4 values.

    - Use the `IMAGE_AND_DAG` type
    - Retrieve the `id`, `imageRepository`, `imageTag`, and `dagsUploadURL`, which you need in the following steps.

2. Log in to Docker with the `ASTRO_API_TOKEN` that you created.

    ```bash

>     docker login images.astronomer.cloud -u cli -p $ASTRO_API_TOKEN

Reviewer: `ASTRO_API_TOKEN` doesn't necessarily need to be configured as an environment variable, what if we just said `<your-astro-api-token>` so that users can configure the API token as they see fit?

---
title: 'Deploy to Astro with the API'
sidebar_label: 'Deploy with API'
id: deploy-with-api
---

> While you can deploy code using the Astro GitHub Integration, the Astro CLI, or by configuring a CI/CD pipeline, you might prefer to use the Astro API to automate deploying code. Using the Astro API is the most straightforward way to deploy code and requires very few additional dependencies, meaning it is compatible with most CI/CD configurations.

Author: ```suggestion
While you can deploy code using the Astro GitHub Integration, the Astro CLI, or by configuring a CI/CD pipeline, you might prefer to use the Astro API to automate deploying code. Using the Astro API is the most straightforward way to deploy code without using the Astro CLI and requires very few additional dependencies, meaning it is compatible with most CI/CD configurations.
```

> While you can deploy code using the Astro GitHub Integration, the Astro CLI, or by configuring a CI/CD pipeline, you might prefer to use the Astro API to automate deploying code. Using the Astro API is the most straightforward way to deploy code without using the Astro CLI and requires very few additional dependencies, meaning it is compatible with most CI/CD configurations.

Reviewer 2: This is good with me

> If the Astro API has access to your Astro project files, you can use the Astro API `deploy` endpoints to complete an image deploy or DAG deploy. You can then implement scripts to automate deploys as a workaround to not using the Astro CLI or Astro GitHub integration.

Reviewer 2: good with me

This best practice guide walks through the steps that are necessary to deploy code to Astro using the Astro API. It also provides example bash scripts that implement these steps for different types of deploy workflows in your own CI/CD pipelines. This example uses Docker to build the image, but you can use a similar tool like Kaniko.

## Feature overview

This guide highlights the following Astro features:

- The Astro API [`Deploy` endpoint](https://www.astronomer.io/api/platform-api-reference/deploy/list-deploys) to create and manage code deploys to an Astro Deployment.

## Prerequisites

These use cases assume you have:

> - Sufficient permissions to deploy code to your Astro Deployments
> - At least one [Astro Deployment](create-deployment.md)

Reviewer: ```suggestion
- Sufficient permissions to deploy code to your Astro Deployments.
- At least one [Astro Deployment](create-deployment.md).
- An [Astro project](cli/develop-project.md) that's accessible from the machine where you're using the Astro API.
```

- A container management tool. The following examples show commands using [Docker](https://www.docker.com/).
- The following values:
    - Your Organization ID - See [List Organizations](https://docs.astronomer.io/docs/api/platform-api-reference/organization/list-organizations)
    - The Deployment ID - See [List Deployments](https://docs.astronomer.io/docs/api/platform-api-reference/deployment/list-deployments)
    - Your Astro API token - See [Create an API token](https://www.astronomer.io/docs/astro/automation-authentication#step-1-create-an-api-token)
    - The path where your Astro project exists.

> ## With DAG-only deploy enabled

Reviewer: ```suggestion
## With DAG-only deploys enabled
```

Reviewer 2: I'm actually not sure what we usually call this

If you have DAG-only deploys enabled, you can create scripts for complete project deploys, DAG-only deploys, or image-only deploys. The following sections include a step by step description of the workflow followed by a bash script that executes the workflow.

### Project deploy

The following steps describe the different actions that the script performs to deploy a complete Astro project.

> 1. Create the `Deploy` API call using the `IMAGE_AND_DAG` type. This action creates an object that represents the intent to deploy code to a Deployment.

Reviewer: ```suggestion
1. Make a `POST` request to the `Deploy` endpoint to create a new `deploy` object. In your call, specify `type` as `IMAGE_AND_DAG`. This action creates an object that represents the intent to deploy code to a Deployment. See the [Astro API documentation](https://www.astronomer.io/docs/api/platform-api-reference/deploy/create-deploy) for request usage and examples.
```

> 1. Create the `Deploy` API call using the `IMAGE_AND_DAG` type. This action creates an object that represents the intent to deploy code to a Deployment.

Reviewer: The user also needs to copy the `DeployID` that's returned in this step! That's a required value for finalizing the deploy later on.

Reviewer 2: @kushalmalani could you confirm?

Reviewer 3: Yes, that is right. You need the deployId to perform FinalizeDeploy

2. After you create the `deploy`, retrieve the `id`, `imageRepository`, `imageTag`, and `dagsUploadURL` values, using the [`GET` deploy API call](https://docs.astronomer.io/docs/api/platform-api-reference/deploy/get-deploy), which you need in the following steps.

3. Log in to Docker or your container management platform, using the Astro API token that you created.

    ```bash

>     docker login images.astronomer.cloud -u cli -p $<your-api-token>

Reviewer 2: @kushalmalani what do you think?

>     docker login images.astronomer.cloud -u cli -p $<your-api-token>

Reviewer 3: Yes, the $ sign is only for the script.

    ```

4. Build the image using the `imageRepository`, `imageTag`, and Astro project path values that you retrieved earlier.

    ```bash

   docker build -t imageRepository:imageTag --platform=linux/amd64 $<astro_project_path>

    ```

5. Push the image using the `imageRepository` and `imageTag` values that you retrieved in Step 2.

    ```bash

    docker push imageRepository:imageTag

    ```

6. Create a tar file of the DAGs folder, where you specify the path where you want to create the tar file.

    ```bash

    tar -cvf <path to create the tar file>/dags.tar "dags"

    ```

>     :::note

Reviewer: ```suggestion
    :::info
```
Pls no more notes 🙏 😭 🙏 😭🙏 😭🙏 😭🙏 😭🙏 😭🙏 😭

>     Make sure to clean up the `dags.tar` file after uploading.

Reviewer: Would it make more sense to have this after the next step?

Reviewer 3: We can do this after the upload is completed

    :::

> 7. Upload the tar file by making a `PUT` call using the `dagsUploadURL` that you retrieved in Step 2. In this call, it is mandatory to pass the following. Then, save the `versionID` from the response header.

Reviewer: - To what endpoint?
- Second sentence seems to be missing some details.
- I don't see the relation between the code fence and the step.

Reviewer 3: the url is generated in the CreateDeploy Endpoint

    ```bash

    --header 'x-ms-blob-type: BlockBlob'

    ```

> 8. Finalize the deploy. See [Finalize the deploy](https://docs.astronomer.io/docs/api/platform-api-reference/deploy/finalize-deploy) for more information about the API request.

Reviewer: ```suggestion
8. Using your `DeployId`, make a request to finalize the deploy. See [Astro API documentation](https://docs.astronomer.io/docs/api/platform-api-reference/deploy/finalize-deploy) for more information about formatting the API request.
```
FYI in case this doesn't get reviewed before I leave, we should really update these docs.

>     - On `Success`, the deploy process has completed. Pass `versionID` in the requested body.

Reviewer: I don't understand this process at all. If the deploy was finalized and succeeded, why do we need to pass any more information?

Reviewer 3: On Step 7, we are uploading the dags tarball to the blob storage, but our services have no idea what versionId was uploaded since that information is only known when the upload is completed. Passing the VersionId in finalizeDeploy is what triggers the deploys process

---
title: 'Deploy to Astro with the API'
sidebar_label: 'Deploy with API'
id: deploy-with-api
---

While you can deploy code using the Astro GitHub integration, the Astro CLI, or by configuring a CI/CD pipeline, there are some environments where you can only use the Astro API to deploy code. This is because using the Astro API has very few dependencies, making it compatible with almost all CI/CD environments and security requirements.

If the Astro API has access to your Astro project files, you can use the Astro API `deploy` endpoints to complete an image deploy or DAG deploy. You can then implement scripts to automate deploys as an alternative to using the Astro CLI or Astro GitHub integration.

> This best practice guide walks through the steps that are necessary to deploy code to Astro using the Astro API. It also provides example bash scripts that implement these steps for different types of deploy workflows in your own CI/CD pipelines. This example uses Docker to build the image, but you can use a similar tool like Kaniko. You can also use a different scripting language, like Python, instead of bash.

Author: ```suggestion
This best practice guide walks through the steps that are necessary to deploy code to Astro using the Astro API. It also provides example bash scripts that implement these steps for different types of deploy workflows in your own CI/CD pipelines. This example uses Docker to build the image, but you can use a similar tool like [kaniko](https://github.com/GoogleContainerTools/kaniko). You can also use a different scripting language, like Python, instead of bash.
```

These use cases assume you have:

- An [API token](deployment-api-tokens.md) with sufficient permissions to deploy code to Astro.
- At least one [Astro Deployment](create-deployment.md).
- An [Astro project](cli/develop-project.md) that's accessible from the machine where you're using the Astro API.
- A container management tool. The following examples use [Docker](https://www.docker.com/).
- The following values:
  - Your Organization ID - See [List Organizations](https://docs.astronomer.io/docs/api/platform-api-reference/organization/list-organizations).
  - The Deployment ID - See [List Deployments](https://docs.astronomer.io/docs/api/platform-api-reference/deployment/list-deployments).
  - Your Astro API token - See [Create an API token](https://www.astronomer.io/docs/astro/automation-authentication#step-1-create-an-api-token).
  - The path where your Astro project exists.

## With DAG-only deploys enabled

If you have DAG-only deploys enabled, you can create scripts for complete project deploys, DAG-only deploys, or image-only deploys. The following sections include a step by step description of the workflow followed by a bash script that executes the workflow.

> ### Project deploy

Reviewer 4: ```suggestion
### Complete project deploy
```

Maybe this helps?

> If you have DAG-only deploys enabled, you can create scripts for complete project deploys, DAG-only deploys, or image-only deploys. The following sections include a step by step description of the workflow followed by a bash script that executes the workflow.

Author: ```suggestion
If you have DAG-only deploys enabled, you can create scripts for complete project deploys, DAG-only deploys, or image-only deploys. The following sections include a step by step description of the workflow followed by a bash script that executes the workflow.

See [Deploy code to Astro](deploy-code.md) for more information about the different ways to update your DAGs and Airflow images.
```

> The following steps describe the different actions that the script performs to deploy a complete Astro project.

Reviewer 4: ```suggestion
The following steps describe the different actions that the script performs to deploy a complete Astro project. A project deploy INCLUDES...
```

I'd very briefly clarify what this means!

1. Make a `POST` request to the `Deploy` endpoint to create a new `deploy` object. In your call, specify `type` as `IMAGE_AND_DAG`. Store the values for `id`, `imageRepository`, `imageTag`, and `dagsUploadURL` that are returned in the response to use in the following steps.

        This action creates an object that represents the intent to deploy code to a Deployment. See the [Astro API documentation](https://www.astronomer.io/docs/api/platform-api-reference/deploy/create-deploy) for request usage and examples.

2. Authenticate to Astronomer's image registry using your Astro API token:

   ```bash

   docker login images.astronomer.cloud -u cli -p <your-api-token>

   ```

> 3. Build the image using the `imageRepository` and `imageTag` values that you retrieved in Step 2, as well as your Astro project path.

Reviewer 3: ```suggestion
3. Build the image using the `imageRepository` and `imageTag` values that you retrieved in Step 1, as well as your Astro project path.
```

   ```bash

   docker build -t imageRepository:imageTag --platform=linux/amd64 <astro_project_path>

   ```

> 4. Push the image using the `imageRepository` and `imageTag` values that you retrieved in Step 2.

Reviewer 3: ```suggestion
4. Push the image using the `imageRepository` and `imageTag` values that you retrieved in Step 1.
```

    FINALIZE_DEPLOY=$(curl --location --request POST "https://api.astronomer.io/platform/v1beta1/organizations/$ORGANIZATION_ID/deployments/$DEPLOYMENT_ID/deploys/$DEPLOY_ID/finalize" \
    --header "X-Astro-Client-Identifier: script" \
    --header "Content-Type: application/json" \
    --header "Authorization: Bearer $ASTRO_API_TOKEN" \
    --data '{"dagTarballVersion": "'$VERSION_ID'"}')

    ID=$(echo $FINALIZE_DEPLOY | jq -r '.id')
    if [[ "$ID" != null ]]; then
            echo -e "\nDeploy is Finalized. Image and DAG changes for deployment $DEPLOYMENT_ID should be live in a few minutes"
            echo "Deployed Image tag: $TAG"
            echo "Deployed DAG Tarball Version: $VERSION_ID"
    else
            MESSAGE=$(echo $FINALIZE_DEPLOY | jq -r '.message')
            if  [[ "$MESSAGE" != null ]]; then
                    echo $MESSAGE
            else
                    echo "Something went wrong. Reach out to astronomer support for assistance"
            fi
    fi

    # Cleanup
    echo -e "\nCleaning up the created tar file from $ASTRO_PROJECT_PATH/dags.tar"
    rm -rf "$ASTRO_PROJECT_PATH/dags.tar"

    ```

</details>

### DAG-only deploy

> 1. Make a `POST` request to the `Deploy` endpoint to create a new `deploy` object. In your call, specify `type` as `DAG_ONLY`. Store the values for `id` and `dagsUploadURL` that are returned in the response to use in the following steps.

Author: ```suggestion
1. Make a `POST` request to the `Deploy` endpoint to create a new `deploy` object. In your call, specify `type` as `DAG_ONLY`. Store the values for `x-ms-version-id` and `dagsUploadURL` that are returned in the response to use in the following steps.
```

> 1. Make a `POST` request to the `Deploy` endpoint to create a new `deploy` object. In your call, specify `type` as `DAG_ONLY`. Store the values for `id` and `dagsUploadURL` that are returned in the response to use in the following steps.

Author: ```suggestion
The following script allows you to update only your DAG files. Refer to [DAG deploys](deploy-dags.md) to learn the details about how Astro deploys DAGs. 

1. Make a `POST` request to the `Deploy` endpoint to create a new `deploy` object. In your call, specify `type` as `DAG_ONLY`. Store the values for `id` and `dagsUploadURL` that are returned in the response to use in the following steps.
```

        This action creates an object that represents the intent to deploy code to a Deployment. See the [Astro API documentation](https://www.astronomer.io/docs/api/platform-api-reference/deploy/create-deploy) for request usage and examples.

2. Create a tar file of your Astro project DAGs folder:

   ```bash

   tar -cvf <path to create the tar file>/dags.tar "dags"

   ```

   :::note

   Make sure to clean up the `dags.tar` file after uploading.

   :::

> 3. Upload the tar file by making a `PUT` call using the `dagsUploadURL` that you retrieved in Step 2. In this call, it is mandatory to pass the `x-ms-blob-type` as `BlockBlob`. Then, save the `versionID` from the response header.

Author: ```suggestion
3. Upload the tar file by making a `PUT` call using the `dagsUploadURL` that you retrieved in Step 2. In this call, it is mandatory to pass the `x-ms-blob-type` as `BlockBlob`. Then, save the `x-ms-version-id` from the response header.
```

4. Using your deploy `id`, make a request to finalize the deploy. See [Astro API documentation](https://docs.astronomer.io/docs/api/platform-api-reference/deploy/finalize-deploy) for more information about formatting the API request.

>    - On `Success`, your DAGs have successfully uploaded and a `versionID` of the DAGs tarball is generated. Pass this `versionID` in the requested body to finish your updates.

Author: ```suggestion
   - On `Success`, your DAGs have successfully uploaded and a `x-ms-version-id` of the DAGs tarball is generated. Pass this `versionID` in the requested body to finish your updates.
```

        FINALIZE_DEPLOY=$(curl --location --request POST "https://api.astronomer.io/platform/v1beta1/organizations/$ORGANIZATION_ID/deployments/$DEPLOYMENT_ID/deploys/$DEPLOY_ID/finalize" \
        --header "X-Astro-Client-Identifier: script" \
        --header "Content-Type: application/json" \
        --header "Authorization: Bearer $ASTRO_API_TOKEN" \
        --data '{"dagTarballVersion": "'$VERSION_ID'"}')

        ID=$(echo $FINALIZE_DEPLOY | jq -r '.id')
        if [[ "$ID" != null ]]; then
                echo -e "\nDeploy is Finalized. DAG changes for deployment $DEPLOYMENT_ID should be live in a few minutes"
                echo "Deployed DAG Tarball Version: $VERSION_ID"
        else
                MESSAGE=$(echo $FINALIZE_DEPLOY | jq -r '.message')
                if  [[ "$MESSAGE" != null ]]; then
                        echo $MESSAGE
                else
                        echo "Something went wrong. Reach out to astronomer support for assistance"
                fi
        fi

        # Cleanup
        echo -e "\nCleaning up the created tar file from $ASTRO_PROJECT_PATH/dags.tar"
        rm -rf "$ASTRO_PROJECT_PATH/dags.tar"

        ```

</details>

> ### Image-only deploy

Author: ```suggestion
### Image-only deploy

The following script allows you to update only your Astro project by building and deploying a new Docker image. Refer to [What happens during a project deploy](deploy-project-image.md#what-happens-during-a-project-deploy) to learn the details about how Astro deploys image updates. 
```

        This action creates an object that represents the intent to deploy code to a Deployment. See the [Astro API documentation](https://www.astronomer.io/docs/api/platform-api-reference/deploy/create-deploy) for request usage and examples.

2. Log in to Docker with your Astro API token.

   ```bash

   docker login images.astronomer.cloud -u cli -p <your_astro_api_token>

   ```

3. Build the image using the `imageRepository` and `imageTag` values that you retrieved in Step 2, as well as your Astro project path.

   ```bash

   docker build -t imageRepository:imageTag --platform=linux/amd64 <astro_project_path>

   ```

4. Push the image using the `imageRepository` and `imageTag` values that you retrieved earlier.

   ```bash

   docker push imageRepository:imageTag

   ```

5. Using your deploy `id`, make a request to finalize the deploy. See [Astro API documentation](https://docs.astronomer.io/docs/api/platform-api-reference/deploy/finalize-deploy) for more information about formatting the API request.

>    - On `Success`, your DAGs have successfully uploaded and a `versionID` of the DAGs tarball is generated. Pass this `versionID` in the requested body to finish your updates.

Author: ```suggestion
   - On `Success`, your DAGs have successfully uploaded and a `x-ms-version-id` of the DAGs tarball is generated. Pass this `x-ms-version-id` in the requested body to finish your updates.
```

  # Step 4: Push image
  echo -e "\nPushing Docker image $REPOSITORY:$TAG to $DEPLOYMENT_ID"
  docker push $REPOSITORY:$TAG

  # Step 5: Finalize Deploy
  FINALIZE_DEPLOY=$(curl --location --request POST "https://api.astronomer.io/platform/v1beta1/organizations/$ORGANIZATION_ID/deployments/$DEPLOYMENT_ID/deploys/$DEPLOY_ID/finalize" \
  --header "X-Astro-Client-Identifier: script" \
  --header "Content-Type: application/json" \
  --header "Authorization: Bearer $ASTRO_API_TOKEN" \
  --data '{}')

  ID=$(echo $FINALIZE_DEPLOY | jq -r '.id')
  if [[ "$ID" != null ]]; then
          echo -e "\nDeploy is Finalized. Image changes for deployment $DEPLOYMENT_ID should be live in a few minutes"
          echo "Deployed Image tag: $TAG"
  else
          MESSAGE=$(echo $FINALIZE_DEPLOY | jq -r '.message')
          if  [[ "$MESSAGE" != null ]]; then
                  echo $MESSAGE
          else
                  echo "Something went wrong. Reach out to astronomer support for assistance"
          fi
  fi
```

</details>

## With DAG-only deploy disabled

You can only use complete project deploys if you have DAG-only deploys disabled. Image-only and DAG-only deploys do not work.

> ### Project deploy

Reviewer 4: Just be consistent if you end up changing the above, if not nothing here

> ### Project deploy

Author: ```suggestion
### Complete project deploy
```

> You can only use complete project deploys if you have DAG-only deploys disabled. Image-only and DAG-only deploys do not work.

Author: ```suggestion
You can only use complete project deploys if you have DAG-only deploys disabled. Image-only and DAG-only deploys do not work.

See [Deploy code to Astro](deploy-code.md) for more information about the different ways to update your DAGs and Airflow images.
```

        echo -e "\nPushing Docker image $REPOSITORY:$TAG to $DEPLOYMENT_ID"
        docker push $REPOSITORY:$TAG

        # Step 6: Finalizing Deploy
        FINALIZE_DEPLOY=$(curl --location --request POST "https://api.astronomer.io/platform/v1beta1/organizations/$ORGANIZATION_ID/deployments/$DEPLOYMENT_ID/deploys/$DEPLOY_ID/finalize" \
        --header "X-Astro-Client-Identifier: script" \
        --header "Content-Type: application/json" \
        --header "Authorization: Bearer $ASTRO_API_TOKEN" \
        --data '{}')

        ID=$(echo $FINALIZE_DEPLOY | jq -r '.id')
        if [[ "$ID" != null ]]; then
                echo -e "\nDeploy is Finalized. Image and DAG changes for deployment $DEPLOYMENT_ID should be live in a few minutes"
                echo "Deployed Image tag: $TAG"
        else
                MESSAGE=$(echo $FINALIZE_DEPLOY | jq -r '.message')
                if  [[ "$MESSAGE" != null ]]; then
                        echo $MESSAGE
                else
                        echo "Something went wrong. Reach out to astronomer support for assistance"
                fi
        fi
        ```

</details>

> ## Trigger deploys when files change

Reviewer 4: Is this actually the absolute recommended way above the H2 process defined above? If so I might make it more explicit in the overview of this doc as a whole, change the title to be `(Recommended)`, or be more explicit below?

Author: I think i might like to have this as an **Example** implementation vs a recommended maybe? @kushalmalani, what do you think?

Reviewer 3: Yes its definitely recommended, since it gives them the best experience with astro.

Author: Okay @paolaperaza + @kushalmalani, overview + the section have been updated

   ```bash

   docker login images.astronomer.cloud -u cli -p <your-api-token>

   ```

3. Build the image using the `imageRepository` and `imageTag` values that you retrieved in Step 1, as well as your Astro project path.

   ```bash

   docker build -t imageRepository:imageTag --platform=linux/amd64 <astro_project_path>

   ```

4. Push the image using the `imageRepository` and `imageTag` values that you retrieved in Step 1.

   ```bash

   docker push imageRepository:imageTag

   ```

5. Create a tar file of your Astro project DAGs folder:

   ```bash

   tar -cvf <path-to-create-tar-file>/dags.tar "dags"

   ```

   :::info

   Make sure to clean up the `dags.tar` file after uploading.

   :::

6. Upload the tar file by making a `PUT` call using the `dagsUploadURL` that you retrieved in Step 1. In this call, it is mandatory to pass the `x-ms-blob-type` as `BlockBlob`. Then, save the `x-ms-version-id` from the response header.

7. Using your deploy `id`, make a request to finalize the deploy. See [Astro API documentation](https://docs.astronomer.io/docs/api/platform-api-reference/deploy/finalize-deploy) for more information about formatting the API request.

>    - On `Success`, your DAGs have successfully uploaded and a `versionID` of the DAGs tarball is generated. Pass this `versionID` in the requested body to finish your updates.

Author: ```suggestion
   - On `Success`, your DAGs have successfully uploaded and a `x-ms-version-id` of the DAGs tarball is generated. Pass this `x-ms-version-id` in the requested body to finish your updates.
```

---
title: 'Deploy to Astro with the API'
sidebar_label: 'Deploy with API'
id: deploy-with-api
---

While you can deploy your Apache Airflow code to Astro using the Astro GitHub integration, the Astro CLI, or by configuring a CI/CD pipeline, your organization might prefer to use the Astro API to deploy code. This is because using the Astro API has very few dependencies, making it compatible with almost all CI/CD environments and security requirements.

> If the Astro API has access to your Astro project files, you can use the `deploy` endpoints in the Astro API to complete either an image deploy or DAG-only deploy. You can then implement scripts to automate deploys as an alternative to using the Astro CLI or the Astro GitHub integration.

Reviewer 3: ```suggestion
If the Astro API has access to your Astro project files, you can use the `deploy` endpoints in the Astro API to complete either an image and dag deploy, image-only deploy or DAG-only deploy. You can then implement scripts to automate deploys as an alternative to using the Astro CLI or the Astro GitHub integration.
```

This best practice guide first walks through the steps that are necessary to deploy code to Astro using the Astro API for three different code deploy methods, a complete project deploy, an image deploy, and a DAGs-only deploy. Then, the guide shows you the recommended way to combine these three automated processes to create a script with conditional logic that can automatically deploy code to Astro, depending on which types of files change in your Astro project.

These examples are bash scripts that use Docker to build the image, but you can use a similar tool like [kaniko](https://github.com/GoogleContainerTools/kaniko). You can also use a different scripting language, like Python, instead of bash.

## Feature overview

This guide highlights the following Astro features:

- The Astro API [`Deploy` endpoint](https://www.astronomer.io/api/platform-api-reference/deploy/list-deploys) to create and manage code deploys to an Astro Deployment.

## Prerequisites

> These use cases assume you have:

Reviewer 4: ```suggestion
This guide assumes that you have:
```

   ```bash

   docker login images.astronomer.cloud -u cli -p <your-api-token>

   ```

3. Build the image using the `imageRepository` and `imageTag` values that you retrieved in Step 1, as well as your Astro project path.

   ```bash

   docker build -t imageRepository:imageTag --platform=linux/amd64 <astro_project_path>

   ```

4. Push the image using the `imageRepository` and `imageTag` values that you retrieved in Step 1.

   ```bash

   docker push imageRepository:imageTag

   ```

5. Create a tar file of your Astro project DAGs folder:

   ```bash

   tar -cvf <path-to-create-tar-file>/dags.tar "dags"

   ```

   :::info

   Make sure to clean up the `dags.tar` file after uploading.

   :::

6. Upload the tar file by making a `PUT` call using the `dagsUploadURL` that you retrieved in Step 1. In this call, it is mandatory to pass the `x-ms-blob-type` as `BlockBlob`. Then, save the `x-ms-version-id` from the response header.

7. Using your deploy `id`, make a request to finalize the deploy. See [Astro API documentation](https://docs.astronomer.io/docs/api/platform-api-reference/deploy/finalize-deploy) for more information about formatting the API request.

>    - On `Success`, your DAGs have successfully uploaded and a `x-ms-version-id` of the DAGs tarball is generated. Pass this `x-ms-version-id` in the requested body to finish your updates.

Reviewer 3: ```suggestion
   - On `Success`, your DAGs have successfully uploaded and a `x-ms-version-id` of the DAGs tarball is generated in the response headers. Pass this `x-ms-version-id` in the requested body to finish your updates.
```

    FINALIZE_DEPLOY=$(curl --location --request POST "https://api.astronomer.io/platform/v1beta1/organizations/$ORGANIZATION_ID/deployments/$DEPLOYMENT_ID/deploys/$DEPLOY_ID/finalize" \
    --header "X-Astro-Client-Identifier: script" \
    --header "Content-Type: application/json" \
    --header "Authorization: Bearer $ASTRO_API_TOKEN" \
    --data '{"dagTarballVersion": "'$VERSION_ID'"}')

    ID=$(echo $FINALIZE_DEPLOY | jq -r '.id')
    if [[ "$ID" != null ]]; then
            echo -e "\nDeploy is Finalized. Image and DAG changes for deployment $DEPLOYMENT_ID should be live in a few minutes"
            echo "Deployed Image tag: $TAG"
            echo "Deployed DAG Tarball Version: $VERSION_ID"
    else
            MESSAGE=$(echo $FINALIZE_DEPLOY | jq -r '.message')
            if  [[ "$MESSAGE" != null ]]; then
                    echo $MESSAGE
            else
                    echo "Something went wrong. Reach out to astronomer support for assistance"
            fi
    fi

    # Cleanup
    echo -e "\nCleaning up the created tar file from $ASTRO_PROJECT_PATH/dags.tar"
    rm -rf "$ASTRO_PROJECT_PATH/dags.tar"

    ```

</details>

### DAG-only deploy

> The following script allows you to update only your DAG files. Refer to [DAG deploys](deploy-dags.md) to learn the details about how Astro deploys DAGs.

Reviewer 4: ```suggestion
The following script allows you to update only your DAG files. Refer to [Deploy DAGs](deploy-dags.md) to learn the details about how Astro deploys DAGs.
```

> 1. Make a `POST` request to the `Deploy` endpoint to create a new `deploy` object. In your call, specify `type` as `DAG_ONLY`. Store the values for `id` and `dagsUploadURL` that are returned in the response to use in the following steps.

Reviewer 4: ```suggestion
1. Make a `POST` request to the `Deploy` endpoint to create a new `deploy` object. In your request, specify `type` as `DAG_ONLY`. Store the values for `id` and `dagsUploadURL` that are returned in the response to use in the following steps.
```

        This action creates an object that represents the intent to deploy code to a Deployment. See the [Astro API documentation](https://www.astronomer.io/docs/api/platform-api-reference/deploy/create-deploy) for request usage and examples.

2. Create a tar file of your Astro project DAGs folder:

   ```bash

   tar -cvf <path to create the tar file>/dags.tar "dags"

   ```

   :::note

   Make sure to clean up the `dags.tar` file after uploading.

   :::

> 3. Upload the tar file by making a `PUT` call using the `dagsUploadURL` that you retrieved in Step 2. In this call, it is mandatory to pass the `x-ms-blob-type` as `BlockBlob`. Then, save the `x-ms-version-id` from the response header.

Reviewer 4: ```suggestion
3. Upload the tar file by making a `PUT` request using the `dagsUploadURL` that you retrieved in Step 2. In this request, it is mandatory to pass the `x-ms-blob-type` as `BlockBlob`. Then, save the `x-ms-version-id` from the response header.
```

        --header "Authorization: Bearer $ASTRO_API_TOKEN" \
        --data '{"dagTarballVersion": "'$VERSION_ID'"}')

        ID=$(echo $FINALIZE_DEPLOY | jq -r '.id')
        if [[ "$ID" != null ]]; then
                echo -e "\nDeploy is Finalized. DAG changes for deployment $DEPLOYMENT_ID should be live in a few minutes"
                echo "Deployed DAG Tarball Version: $VERSION_ID"
        else
                MESSAGE=$(echo $FINALIZE_DEPLOY | jq -r '.message')
                if  [[ "$MESSAGE" != null ]]; then
                        echo $MESSAGE
                else
                        echo "Something went wrong. Reach out to astronomer support for assistance"
                fi
        fi

        # Cleanup
        echo -e "\nCleaning up the created tar file from $ASTRO_PROJECT_PATH/dags.tar"
        rm -rf "$ASTRO_PROJECT_PATH/dags.tar"

        ```

</details>

### Image-only deploy

The following script allows you to update only your Astro project by building and deploying a new Docker image. Refer to [What happens during a project deploy](deploy-project-image.md#what-happens-during-a-project-deploy) to learn the details about how Astro deploys image updates.

> 1.  Make a `POST` request to the `Deploy` endpoint to create a new `deploy` object. In your call, specify `type` as `IMAGE`. Store the value for the `DeployID` that is returned. This action creates an object that represents the intent to deploy code to a Deployment.  See the [Astro API documentation](https://www.astronomer.io/docs/api/platform-api-reference/deploy/create-deploy) for request usage and examples.

Reviewer 4: ```suggestion
1.  Make a `POST` request to the `Deploy` endpoint to create a new `deploy` object. In your request, specify `type` as `IMAGE`. Store the value for the `DeployID` that is returned. This action creates an object that represents the intent to deploy code to a Deployment.  See the [Astro API documentation](https://www.astronomer.io/docs/api/platform-api-reference/deploy/create-deploy) for request usage and examples.
```

> 1. Make a `POST` request to the `Deploy` endpoint to create a new `deploy` object. In your call, specify `type` as `IMAGE`. Store the values for `id`, `imageRepository`, and `imageTag` that are returned in the response to use in the following steps.

Reviewer 4: ❌  Is this a duplicate step 1?

Author: Sure is!

> 1. Make a `POST` request to the `Deploy` endpoint to create a new `deploy` object. In your call, specify `type` as `IMAGE`. Store the values for `id`, `imageRepository`, and `imageTag` that are returned in the response to use in the following steps.
> 
>         This action creates an object that represents the intent to deploy code to a Deployment. See the [Astro API documentation](https://www.astronomer.io/docs/api/platform-api-reference/deploy/create-deploy) for request usage and examples.

Author: ```suggestion
```

          echo -e "\nDeploy is Finalized. Image changes for deployment $DEPLOYMENT_ID should be live in a few minutes"
          echo "Deployed Image tag: $TAG"
  else
          MESSAGE=$(echo $FINALIZE_DEPLOY | jq -r '.message')
          if  [[ "$MESSAGE" != null ]]; then
                  echo $MESSAGE
          else
                  echo "Something went wrong. Reach out to astronomer support for assistance"
          fi
  fi
```

</details>

## With DAG-only deploy disabled

You can only use complete project deploys if you have DAG-only deploys disabled. Image-only and DAG-only deploys do not work.

See [Deploy code to Astro](deploy-code.md) for more information about the different ways to update your DAGs and Airflow images.

### Complete project deploy

1. Create the `Deploy` API call using the `IMAGE_AND_DAG` type. This action creates an object that represents the intent to deploy code to a Deployment.

2. After you create the `deploy`, retrieve the `id`, `imageRepository`, and `imageTag` values using the [`GET` deploy API call](https://docs.astronomer.io/docs/api/platform-api-reference/deploy/get-deploy), which you need in the following steps.

> 3. Log in to Docker

Reviewer 4: ```suggestion
3. Log in to Docker with your Astro API token.
```

        echo -e "\nPushing Docker image $REPOSITORY:$TAG to $DEPLOYMENT_ID"
        docker push $REPOSITORY:$TAG

        # Step 6: Finalizing Deploy
        FINALIZE_DEPLOY=$(curl --location --request POST "https://api.astronomer.io/platform/v1beta1/organizations/$ORGANIZATION_ID/deployments/$DEPLOYMENT_ID/deploys/$DEPLOY_ID/finalize" \
        --header "X-Astro-Client-Identifier: script" \
        --header "Content-Type: application/json" \
        --header "Authorization: Bearer $ASTRO_API_TOKEN" \
        --data '{}')

        ID=$(echo $FINALIZE_DEPLOY | jq -r '.id')
        if [[ "$ID" != null ]]; then
                echo -e "\nDeploy is Finalized. Image and DAG changes for deployment $DEPLOYMENT_ID should be live in a few minutes"
                echo "Deployed Image tag: $TAG"
        else
                MESSAGE=$(echo $FINALIZE_DEPLOY | jq -r '.message')
                if  [[ "$MESSAGE" != null ]]; then
                        echo $MESSAGE
                else
                        echo "Something went wrong. Reach out to astronomer support for assistance"
                fi
        fi
        ```

</details>

> ## (Recommended) Trigger deploys when files change

Reviewer 4: Can we put this as the first H2 if it's recommended?

Author: So i think the *Recommended* part is misleading, it's not recommended necessarily above the other three options, it's the recommended way to *combine* the three.

Author: clarified in expanded overview and intro section

> ## (Recommended) Trigger deploys when files change

Reviewer 4: ```suggestion
## (Recommended) Dynamic deploys based on files changed
```

In addition to manually triggering project, DAG, and image deploys, you can create scripts that trigger specific code deploys depending on whether there have been changes to DAGs or files used to build the project image.

The following code example shows the recommended steps for you to use to create a bash script that triggers different code deploys depending on the files changed in your project. This example demonstrates how to combine all three types of deploys and the conditional logic for when to deploy each.

:::info

> This recommended process requires that you enable DAG-only deploys.

Reviewer 3: Can we link the page to enable DAG deploys

> This recommended process requires that you enable DAG-only deploys.

Author: ```suggestion
This recommended process requires that you enable [DAG-only deploys](deploy-dags.md).
```

:::

> ### Recommended workflow

Reviewer 4: ```suggestion
### Dynamic deploys workflow
```

Reviewer 4: I feel like we're missing just 1 intro sentence here. I read the steps and I'm a little bit like -- "Am I supposed to be doing anything?" It sounds like this is just explaining what the script does
